src/modules/interface/controller_mellinger.h

```diff
@@ -27,12 +27,82 @@
 #define __CONTROLLER_MELLINGER_H__
 
 #include "stabilizer_types.h"
+#include "math3d.h"
 
-void controllerMellingerInit(void);
-bool controllerMellingerTest(void);
-void controllerMellinger(control_t *control, const setpoint_t *setpoint,
+typedef struct {
+    float mass;
+    float massThrust;
+
+    // XY Position PID
+    float kp_xy;      // P
+    float kd_xy;      // D
+    float ki_xy;      // I
+    float i_range_xy;
+
+    // Z Position
+    float kp_z;       // P
+    float kd_z;       // D
+    float ki_z;       // I
+    float i_range_z;
+
+    // Attitude
+    float kR_xy;      // P
+    float kw_xy;      // D
+    float ki_m_xy;    // I
+    float i_range_m_xy;
+
+    // Yaw
+    float kR_z;       // P
+    float kw_z;       // D
+    float ki_m_z;     // I
+    float i_range_m_z;
+
+    // roll and pitch angular velocity
+    float kd_omega_rp; // D
+
+    // Helper variables
+    float i_error_x;
+    float i_error_y;
+    float i_error_z;
+
+    float prev_omega_roll;
+    float prev_omega_pitch;
+    float prev_setpoint_omega_roll;
+    float prev_setpoint_omega_pitch;
+
+    float i_error_m_x;
+    float i_error_m_y;
+    float i_error_m_z;
+
+    // Logging variables
+    struct vec z_axis_desired;
+
+    float cmd_thrust;
+    float cmd_roll;
+    float cmd_pitch;
+    float cmd_yaw;
+    float r_roll;
+    float r_pitch;
+    float r_yaw;
+    float accelz;
+} controllerMellinger_t;
+
+void controllerMellingerInit(controllerMellinger_t* self);
+bool controllerMellingerTest(controllerMellinger_t* self);
+void controllerMellinger(controllerMellinger_t* self, control_t *control, const setpoint_t *setpoint,
+                                         const sensorData_t *sensors,
+                                         const state_t *state,
+                                         const uint32_t tick);
+
+#ifdef CRAZYFLIE_FW
```

Comment:
I actually never noticed the CRAZYFLIE_FW define before. Is this to separate functionality for the python bindings build and the firmware build? Will there be an build error if you don't use it?

Reply:
I copied this logic from collision_avoidance.c - might have been introduced there by @jpreiss. We want it here to avoid that these functions end up in the bindings, since they are "internal". A cleaner way could be to have separate files, but that makes code navigation more annoying.

Reply:
I considered both choices. IIRC, I didn't have particularly strong motivation for choosing one file with `#ifdef` over two files. I probably just did it to reduce the number of files. I don't see any discussion in https://github.com/bitcraze/crazyflie-firmware/issues/567 either.